### DISCUSSION TITLE
Deployment breaks between npm 8 and npm 9 due to local symlinks

### DISCUSSION THREAD

**Discussion by sffc** [npm/feedback#864] [CLI]
Disclaimer: I've read about the fairly new feature of npm workspaces to handle multi-project repositories, and I should probably migrate over to them, but I am nonetheless filing this issue to highlight a regression in behavior between npm versions.

I have done my research and have not been able to find another discussion or issue on this subject.

## Description

I have the following directory structure:

- `front`
    - `package.json`
- `shared`
    - `package.json`
- `config.json`

Inside [`front/package.json`](https://github.com/octave-online/octave-online-server/blob/8be12d8ce38a62c1474e4dacf92897b6976962ab/front/package.json#L17), I have a dependency on the `shared` project, like so:

```json
"@oo/shared": "file:../shared"
```

## Behavior Difference

When I run `npm install` inside `front`, npm version 8 creates a symlink:

```bash
$ ls -l node_modules/@oo/
shared -> ../../../shared
```

However, npm version 9 seems to _copy_ the package. There is no symlink.

This breaks my deployment because the `shared` project reads the `config.json` file from the parent directory. Previously, with the symlink, this worked fine, but now that the package is copied over, this no longer works. It's probably bad style for code to read relative paths like this, but _it used to work_, and now _it doesn't work_.

## Next Steps

For the time being, I am working around this issue by pinning `npm@8` in the deploy script. However, I would like clarity on:

1. Whether this is a bug or a feature "working as intended" in npm 9
2. What is the best repository and package structure to be able to reliably read a config.json file saved in the root of the repository

**Comment by ljharb** [comment]
Add `install-links=false` to `.npmrc`, and i think that will restore the npm 8 behavior for you.

**Reply by sffc** [reply]
Can confirm that this appears to work!

**Comment by wraithgar** [comment]
1. Yes this is intentional. In npm 9 the `install-links` value was changed to default to true.
2. npm auto-discovers the package root already (that's how it knows where to read package.json even if you're in a subdirectory) and exports that value to `$npm_config_local_prefix`.  Use that environment variable to navigate from the project root.
